docs/source/data/fetching-data.mdx

```diff
@@ -1,6 +1,298 @@
+
+Your data source class can be as straightforward or complex as you need it to be. You know what data your server needs, and you can let that be the guide for the methods your class includes.
+
+Below is an example of a data source class that connects to a database storing reservations:
+
+<MultiCodeBlock>
+
+```ts title="reservations.ts"
+export class ReservationsDataSource {
+  private dbConnection;
+  private token;
+  private user;
+
+  constructor(options: { token: string }) {
+    this.dbConnection = this.initializeDBConnection();
+    this.token = options.token;
+  }
+
+  async initializeDBConnection() {
+    // set up our database details, instantiate our connection,
+    // and return that database connection
+    return dbConnection;
+  }
+
+  async getUser() {
+    if (!this.user) {
+      // store the user, lookup by token
+      this.user = await this.dbConnection.User.findByToken(this.token);
+    }
+    return this.user;
+  }
+
+  async getReservation(reservationId) {
+    const user = await this.getUser(this.token);
+    if (user) {
+      return await this.dbConnection.Reservation.findByPk(reservationId);
+    } else {
+      // handle invalid user
+    }
+  }
+
+  //... more methods for finding and creating reservations
+}
+```
+
+</MultiCodeBlock>
+
+> Apollo's [`RestDataSource`](https://github.com/apollographql/datasource-rest/blob/main/src/RESTDataSource.ts) class is a built-out example of how a data source class can handle caching, deduplication, and errors.
+
+### Batching and caching
+
+If you want to add batching, deduplication, or caching to your data source class, we recommend using the [DataLoader package](https://github.com/graphql/dataloader). Using a package like DataLoader is particularly helpful for solving the [infamous N+1 query problem](https://shopify.engineering/solving-the-n-1-problem-for-graphql-through-batching).
+
+DataLoader provides a memoization cache, which avoids loading the same object multiple times during a single GraphQL request ([much like one of `RESTDataSource`'s caching layers](./fetching-rest#get-requests-and-responses)). It also combines loads during a single event loop tick into a batched request that fetches multiple objects at once.
+
+DataLoader instances are pre-request, so if you use a DataLoader in your data source, ensure you **create a new instance of that class with every request**/ :
+
+```ts
+import DataLoader from 'dataloader';
+
+class ProductsDataSource {
+  private dbConnection;
+
+  constructor() {
+    this.dbConnection = this.initializeDBConnection();
+  }
+
+  async initializeDBConnection() {
+    // set up our database details, instantiate our connection,
+    // and return that database connection
+    return dbConnection;
+  }
+
+  private batchProducts = new DataLoader(async (ids) => {
+    const progressList = await db.fetchAllKeys(ids);
+    return ids.map((id) => progressList.find((progress) => progress.id === id));
+  });
+
+  async getProgressFor(id) {
```

Comment:
not really sure what "progress" is in this example

Reply:
I updated this to be more "product"-centric. How does this feel, instead?

```ts
class ProductsDataSource {
  private dbConnection;

  private batchProducts = new DataLoader(async (ids) => {
    const productList = await db.fetchAllKeys(ids);
    return ids.map((id) => productList.find((product) => product.id === id));
  });

  async getProductFor(id) {
    return this.batchProducts.load(id);
  }
}
```